src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/DefineConstantsCSharpValueProvider.cs

```diff
@@ -76,9 +76,18 @@ public override async Task<bool> IsValueDefinedInContextAsync(string propertyNam
         IEnumerable<string> innerConstants =
             ParseDefinedConstantsFromUnevaluatedValue(await defaultProperties.GetUnevaluatedPropertyValueAsync(ConfiguredBrowseObject.DefineConstantsProperty) ?? string.Empty);
 
-        IEnumerable<string> constantsToWrite = KeyValuePairListEncoding.Parse(unevaluatedPropertyValue)
+        // we receive a comma-separated list, we should convert it to a semicolon-separated list
+        // because each item may have multiple values separated by semicolons
+        // ie "A,B;C" should be converted to "A;B;C"
+        unevaluatedPropertyValue = KeyValuePairListEncoding.Format(
+            KeyValuePairListEncoding.Parse(unevaluatedPropertyValue, separator: ','),
+            separator: ";");
```

Comment:
Why can't we just replace `,` with `;`? If there are cases where the `,` would be escaped, please cover them in `KeyValuePairListEncodingTests`. Otherwise I don't see what this round trip is adding. If needed, please add more info in the comment (and more tests if needed).

Would be great to have a unit test of this method given there was a bug in it.

Reply:
It's useful to keep the distinction between raw values, because then you can, for example, notify the caller that a value was transformed into other values and should be removed from the selection list

ie, user enters Constant2;Constant3 in the control, and has already selected Constant1. We'd like to remove `Constant2;Constant3` from the selected options in the control, but could not do that if the raw input to this method was `Constant1;Constant2;Constant3`

Reply:
Added more unit tests of the property provider

Reply:
Is there an example of where:

```c#
KeyValuePairListEncoding.Format(
    KeyValuePairListEncoding.Parse(unevaluatedPropertyValue, separator: ','),
    separator: ";")
```

is not equivalent to

```c#
unevaluatedPropertyValue.Replace(",", ";")
```

?

I couldn't think of one. Performing a string replace will be more efficient.

Reply:
Based on [this comment](https://github.com/dotnet/project-system/pull/9612#discussion_r1881355814), I made it so that we are only receiving a comma-separated list, the values cannot contain semicolons of their own.